sound/soc/sof/control.c

```diff
@@ -342,21 +345,41 @@ int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
 	struct snd_sof_dev *sdev = scontrol->sdev;
 	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
 	unsigned int tag = 0;
+	int max_length = SOF_IPC_MSG_MAX_SIZE -
+		sizeof(const struct sof_ipc_ctrl_data) -
+		sizeof(const struct sof_abi_hdr);
 
 	pm_runtime_get_sync(sdev->dev);
 
+	/* Decrement size for room for ext bytes header and find out
+	 * the limit from ext bytes size from topology and SOF IPC
+	 * size limit.
+	 */
+	size -= 2 * sizeof(u32);
+	if (max_length > be->max)
+		max_length = be->max;
+
+	dev_dbg(sdev->dev, "request size minus header is %d\n", size);
+	dev_dbg(sdev->dev, "max size is %d\n", max_length);
+
+	/* Limit size to prevent read of kernel data */
+	if (size > max_length) {
+		size = max_length;
+		dev_dbg(sdev->dev, "warning: the copy length was limited\n");
+	}
+
 	dev_dbg(sdev->dev, "getting data and command is %d\n", scontrol->cmd);
 	/* get all the mixer data from DSP */
 	snd_sof_ipc_get_comp_data(sdev->ipc, scontrol, SOF_IPC_COMP_GET_DATA,
 				  SOF_CTRL_TYPE_DATA_GET, scontrol->cmd);
 
-	/* TODO: replace 252 with actual size */
 	if (copy_to_user(binary_data, &tag, sizeof(u32)))
 		return -EFAULT;
+
 	if (copy_to_user(binary_data + 1, &size, sizeof(u32)))
 		return -EFAULT;
-	if (copy_to_user(binary_data + 2, cdata->data->data,
-			 SOF_IPC_MSG_MAX_SIZE - sizeof(struct sof_ipc_ctrl_data)))
+
+	if (copy_to_user(binary_data + 2, cdata->data->data, size))
```

Comment:
@singalsu should this be "size" bytes or max_length bytes or the size sent by the FW?

Reply:
The design problem with ASoC get is that the size should originate from firmware or at least have possibility to return the actual bytes read. Now the request is for the topology specified length but since it varies the request is always the max length. I used as example for this fix the similar ext bytes get function for Skylake.

But it should be acceptable for the user space to get a max length response and then use the uapi headers to get the relevant data. In EQ case the first uint32_t word after the user copy header is the length to use.

So, I'd keep this as is in the patch.